GraphicsToolsProject/Plugins/GraphicsTools/Source/GraphicsTools/Public/GTWorldSubsystem.h

```diff
@@ -22,10 +21,22 @@ class GRAPHICSTOOLS_API UGTWorldSubsystem : public UWorldSubsystem
 
 public:
 	/** List of all DirectionalLights within a world. */
-	TArray<UGTDirectionalLightComponent*> DirectionalLights;
+	TArray<UGTSceneComponent*> DirectionalLights;
 
 	/** List of all ProximityLights within a world. */
-	TArray<UGTProximityLightComponent*> ProximityLights;
+	TArray<UGTSceneComponent*> ProximityLights;
+
+	/** List of all ClippingPlanes within a world. */
+	TArray<UGTSceneComponent*> ClippingPlanes;
+
+	/** List of all ClippingSpheres within a world. */
+	TArray<UGTSceneComponent*> ClippingSpheres;
+
+	/** List of all ClippingBoxes within a world. */
+	TArray<UGTSceneComponent*> ClippingBoxes;
+
+	/** List of all ClippingCones within a world. */
+	TArray<UGTSceneComponent*> ClippingCones;
```

Comment:
If you are using these arrays just to assign indices to each component a possible alternative to explore would be to:
1. Have a map here with an entry for each component type containing a list of free indices and provide Acquire/ReleaseIndex() methods that receive the component type (use UE reflection info).
2. Call Acquire/ReleaseIndex() from GTSceneComponent with the dynamic type of the component and store the index.
There may be drawbacks to that, I haven't thought it through that much.

Reply:
You are correct, these really just boil down to arrays to determine indices. 😁 I'm using them mostly as a queue to ensure order. 

For example if we have 5 lights in our world and we only support 3 at I time. If a user deletes the 2nd light I want lights 1, 3, and 4 to now be active and have light 5 be "on deck" if any predecessors are deleted. Which you highlight as a possible issue with inconsecutive indices in a acquire/release scheme.

Trying to think of simple solutions based on your suggestion. I could iterate over all of the UObjects in the level and hope they are kept ordered. But, that sounds slow. I could also make this API private with friends? But, maybe users will want to iterate over these things. Open to ideas!

Reply:
I see. I can't think of a better solution than keeping lists of components of a given type per world.